api/jpo-conflictvisualizer-api/src/main/java/us/dot/its/jpo/ode/api/services/ReportService.java

```diff
@@ -97,66 +115,274 @@ public class ReportService {
+
+                LaneDirectionOfTravelReportData reportData = new LaneDirectionOfTravelReportData();
+                reportData.setTimestamp(minute);
+                reportData.setLaneID(group.getLaneID());
+                reportData.setSegmentID(group.getSegmentID());
+                reportData.setHeadingDelta(group.getMedianHeading() - group.getExpectedHeading());
+                reportData.setMedianCenterlineDistance(group.getMedianCenterlineDistance());
+
+                groupedData
+                    .computeIfAbsent(group.getLaneID(), k -> new HashMap<>())
+                    .computeIfAbsent(group.getSegmentID(), k -> new ArrayList<>())
+                    .add(reportData);
+            }
+        }
+
+        // Aggregate data by minute for each lane ID and segment ID
+        for (Map.Entry<Integer, Map<Integer, List<LaneDirectionOfTravelReportData>>> laneEntry : groupedData.entrySet()) {
+            int laneID = laneEntry.getKey();
+            for (Map.Entry<Integer, List<LaneDirectionOfTravelReportData>> segmentEntry : laneEntry.getValue().entrySet()) {
+                int segmentID = segmentEntry.getKey();
+                Map<Long, List<LaneDirectionOfTravelReportData>> aggregatedData = segmentEntry.getValue().stream()
+                    .collect(Collectors.groupingBy(LaneDirectionOfTravelReportData::getTimestamp));
+
+                for (Map.Entry<Long, List<LaneDirectionOfTravelReportData>> minuteEntry : aggregatedData.entrySet()) {
+                    long minute = minuteEntry.getKey();
+                    List<LaneDirectionOfTravelReportData> minuteData = minuteEntry.getValue();
+
+                    double averageHeadingDelta = minuteData.stream()
+                        .mapToDouble(LaneDirectionOfTravelReportData::getHeadingDelta)
+                        .average()
+                        .orElse(0.0);
+
+                    double averageCenterlineDistance = minuteData.stream()
+                        .mapToDouble(LaneDirectionOfTravelReportData::getMedianCenterlineDistance)
+                        .average()
+                        .orElse(0.0);
+
+                    LaneDirectionOfTravelReportData aggregatedReportData = new LaneDirectionOfTravelReportData();
+                    aggregatedReportData.setTimestamp(minute);
+                    aggregatedReportData.setLaneID(laneID);
+                    aggregatedReportData.setSegmentID(segmentID);
+                    aggregatedReportData.setHeadingDelta(averageHeadingDelta);
+                    aggregatedReportData.setMedianCenterlineDistance(averageCenterlineDistance);
+
+                    reportDataList.add(aggregatedReportData);
+                }
+            }
+        }
+
+        return reportDataList;
+    }
 
-        String reportName = "CmReport_"+ intersectionID + "_" + roadRegulatorID + "_" + startTime + "_" + endTime;
+    private List<StopLineStopReportData> aggregateStopLineStopEvents(List<StopLineStopEvent> events) {

```

Comment:
There are a lot of these aggregation functions here. Is this the best place for these? Or should they be moved into their own class? Would it make sense to split them up and put each one alongside the data structure it works with? I don't know one way or another, just wanted to bring this up for discussion.

Reply:
Sorry, I've been laid low by the flu for a few days, and was finally lucid enough this afternoon to implement these fixes. I thought your suggestion to reorganize them with their respective structures sounded tidy, so that's what I included in the new commit, which also resolved your other comments. Let me know if you would like further discussion.